### DISCUSSION TITLE
Loading Web Assembly Binaries from a given Path and using AssemblyScript

### DISCUSSION THREAD

**Discussion by SRHMan** [nesbox/TIC-80#1909] [General]
How to I load a binary .wasm file in TIC-80 from a given path?  Right now, I have to copy the binary into my %appdata% com.nesbox.tic folder in order for it to find it and that makes the build process more difficult.  Also I get an "unknown option" when using the -code argument.  Did something change?
(On pro version). 

Also, I wrote a simple Test in AssebemblyScript and got it to build: 

declare function print(s: string,x: i32,y: i32): void;
export function TIC():void
{
	 print("this is a test",10,10);	     
}

Error: WASM must export a TIC function.
Looking at the text file of the compiled assembly, it looks like
it is exported.  Has anyone tried using AssebmblyScript for tic80?

**Comment by nesbox** [comment]
Try this command `tic80 --fs . --cmd 'load game.tic & import binary cart.wasm & save & exit'` to import .wasm from the working directory.
Don't know about AssebmblyScript, maybe @joshgoebel knows?

**Comment by joshgoebel** [comment]
> Looking at the text file of the compiled assembly,

But you're importing the binary, right?

> Error: WASM must export a TIC function.

I worry if somehow you haven't done the import at all... pretty sure you also get this error if the `BINARY` segment is entirely empty... There is nothing magic about AssemblyScript vs C vs Zig once they are compiled into a WASM binary - if the `TIC` symbol is exported it will be run - so either your final binary isn't exporting it - or it isn't being importing into TIC-80...

You might look at the WASM-4 console examples though and if you wanted to port over their AsssemblyScript stuff so we "officially" support it (by providing an easy to use build template/example) that'd be awesome.

**Comment by joshgoebel** [comment]
> There is nothing magic about 

Well actually getting memory in the right segments matter, which is why I refer you to the WASM-4 examples... but that would be a problem you'd run into AFTER you get past the "can't find TIC" error...

**Comment by SRHMan** [comment]
Thanks for the tip nesbox, that works pretty good.  I was not putting the '&' between commands.  I'll keep experimenting with assembly script and when I get something basic working, I'll post it here.  I look at the textual output of assemblyscript and see if it is actually in the right segement.

**Comment by SRHMan** [comment]
I don't know, but seems like webassembly is more of a complicated beast than I anticipated.  I was hoping to be able to create web assembly binaries without actually have to know web assembly itself. But anyway here is an example output from the compiler.  It doesn't seem to have that error I had previously, but as soon as I try to declare a function it gives me that error.  

Here is the .ts file:

declare function cls(n: i32): void
export function TIC(): void
{
   cls(0);
}

and here is the output from the compiler:

(module
 (type $i32_=>_none (func (param i32)))
 (type $none_=>_none (func))
 (import "tictest" "cls" (func $src/tictest/cls (param i32)))
 (memory $0 0)
 (export "TIC" (func $src/tictest/TIC))
 (export "memory" (memory $0))
 (func $src/tictest/TIC
  i32.const 0
  call $src/tictest/cls
 )
)

**Comment by joshgoebel** [comment]
That looks sort of ok, but the `$src/tictest` stuff and `tictest` imports look weird... TIC-80 does not export any `tictest` symbols... I'd still recommend what I suggested earlier:

- Start with WASM-4 (which has great AS support), not TIC-80
  - https://github.com/aduros/wasm4/tree/main/cli/assets/templates/assemblyscript
- Get a tiny example working on WASM-4
- Port over your example and build template to TIC-80 (see instructions in `tic80/templates`)
- Now use all the knowledge you've learned to try again to get it working on TIC-80


You'll need to adjust the memory config slightly for TIC-80:

```
    "initialMemory": 4,
    "maximumMemory": 4,
    "memoryBase":98304,
```
